File: app/src/main/java/korablique/recipecalculator/model/FoodstuffsTopList.java
package korablique.recipecalculator.model;

import org.joda.time.DateTime;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import korablique.recipecalculator.base.Callback;
import korablique.recipecalculator.base.TimeProvider;
import korablique.recipecalculator.database.DatabaseWorker;
import korablique.recipecalculator.database.HistoryWorker;

@Singleton
public class FoodstuffsTopList {
    private static final int TOP_LIMIT = 5;
    private final HistoryWorker historyWorker;
    private final TimeProvider timeProvider;
    private final List<Observer> observers = new ArrayList<>();

    public interface Observer {
        void onFoodstuffsTopPossiblyChanged();
    }

    @Inject
    public FoodstuffsTopList(
            HistoryWorker historyWorker,
            TimeProvider timeProvider) {
        this.historyWorker = historyWorker;
        this.timeProvider = timeProvider;
        historyWorker.addObserver(new HistoryWorker.Observer() {
            @Override
            public void onHistoryChange() {
                for (Observer observer : observers) {
                    observer.onFoodstuffsTopPossiblyChanged();
                }
            }
        });
    }

    public void addObserver(Observer observer) {
        observers.add(observer);
    }

    public void removeObserver(Observer observer) {
        observers.remove(observer);
    }

    public void getTopList(Callback<List<Foodstuff>> resultCallback) {
        requestTopFoodstuffs(TOP_LIMIT, (foodstuffs) -> {
            resultCallback.onResult(foodstuffs);
        });
    }

    private void requestTopFoodstuffs(int limit, Callback<List<Foodstuff>> callback) {
        // Сначала делаем запросы в БД, в коллбеках сохраняем результаты,
        // а затем уже добавляем в адаптеры элементы.
        // Это нужно для того, чтобы элементы на экране загружались все сразу
        List<Foodstuff> foodstuffs = new ArrayList<>(); // это продукты за период
        DateTime now = timeProvider.now();
        DateTime monthAgo = now.minusMonths(1).withTimeAtStartOfDay();
        DateTime tomorrowMidnight = now.plusDays(1).withTimeAtStartOfDay();
        historyWorker.requestListedFoodstuffsFromHistoryForPeriod(
                monthAgo.toDate().getTime(),
                tomorrowMidnight.toDate().getTime(),
                (listedFoodstuffs) -> {
                    foodstuffs.addAll(listedFoodstuffs);
                    List<PopularProductsUtils.FoodstuffFrequency> topList = PopularProductsUtils.getTop(foodstuffs); // это топ из них
                    List<Foodstuff> topFoodstuffs = new ArrayList<>();
                    for (int index = 0; index < topList.size() && index < limit; index++) {
                        topFoodstuffs.add(topList.get(index).getFoodstuff());
                    }
                    callback.onResult(topFoodstuffs);
                });
    }
}

File: app/src/main/java/korablique/recipecalculator/model/FoodstuffsTopList.kt
package korablique.recipecalculator.model

import org.joda.time.DateTime

import java.util.ArrayList

import javax.inject.Inject
import javax.inject.Singleton

import io.reactivex.Observable
import io.reactivex.Single
import korablique.recipecalculator.base.TimeProvider
import korablique.recipecalculator.database.FoodstuffsList
import korablique.recipecalculator.database.HistoryWorker

@Singleton
open class FoodstuffsTopList @Inject
constructor(
        private val historyWorker: HistoryWorker,
        foodstuffsList: FoodstuffsList,
        private val timeProvider: TimeProvider) {
    private val observers = mutableListOf<Observer>()
    private var monthTop: Observable<Foodstuff> = Observable.empty()
    private var weekTop: Observable<Foodstuff> = Observable.empty()

    interface Observer {
        fun onFoodstuffsTopPossiblyChanged()
    }

    init {
        updateCache()
        historyWorker.addHistoryChangeObserver {
            updateCache()
            observers.forEach { it.onFoodstuffsTopPossiblyChanged() }
        }
        foodstuffsList.addObserver(object : FoodstuffsList.Observer {
            override fun onFoodstuffEdited(edited: Foodstuff) {
                // Заменяем в закешированном списке отредактированный
                // продукт на его отредактированную версию
                monthTop = monthTop
                        .map { foodstuff ->
                            if (foodstuff.id == edited.id) {
                                edited
                            } else {
                                foodstuff
                            }
                        }
                weekTop = weekTop
                        .map { foodstuff ->
                            if (foodstuff.id == edited.id) {
                                edited
                            } else {
                                foodstuff
                            }
                        }
                observers.forEach { it.onFoodstuffsTopPossiblyChanged() }
            }

            override fun onFoodstuffDeleted(deleted: Foodstuff) {
                monthTop = monthTop.filter { foodstuff -> foodstuff.id != deleted.id }
                weekTop = weekTop.filter { foodstuff -> foodstuff.id != deleted.id }
                observers.forEach { it.onFoodstuffsTopPossiblyChanged() }
            }
        })
    }

    private fun updateCache() {
        val now = timeProvider.now()
        val monthAgo = now.minusMonths(1).withTimeAtStartOfDay()
        val weekAgo = now.minusWeeks(1).withTimeAtStartOfDay()
        val tomorrowMidnight = now.plusDays(1).withTimeAtStartOfDay()
        this.monthTop = requestTopFoodstuffs(monthAgo, tomorrowMidnight).cache()
        this.monthTop.subscribe()
        this.weekTop = requestTopFoodstuffs(weekAgo, tomorrowMidnight).cache()
        this.weekTop.subscribe()
    }

    fun addObserver(observer: Observer) {
        observers.add(observer)
    }

    fun removeObserver(observer: Observer) {
        observers.remove(observer)
    }

    /**
     * Функция возвращает рассчитанный топ продуктов за месяц.
     * Топ продуктов меняется при каждом обновлении таблицы с историей в БД, поэтому
     * если таблица недавно менялась, то возможно топ ещё не вычислен и
     * операция будет не моментальна.
     * Если история менялась давно, то топ уже вычислен и подписка на него отдаст результат
     * моментально.
     */
    open fun getMonthTop(): Single<List<Foodstuff>> {
        return monthTop.toList()
    }

    /**
     * @see getMonthTop
     */
    open fun getWeekTop(): Single<List<Foodstuff>> {
        return weekTop.toList()
    }

    private fun requestTopFoodstuffs(from: DateTime, to: DateTime): Observable<Foodstuff> {
        val from = from.toDate().time
        val to = to.toDate().time
        return historyWorker.requestListedFoodstuffsFromHistoryForPeriod(from, to)
                .toList()
                .map { foodstuffs ->
                    // Преобразуем набор продуктов из истории в топ, а затем обратно
                    // в список продуктов.
                    val topList = PopularProductsUtils.getTop(foodstuffs)
                    topList.map { it.foodstuff }
                }
                .toObservable()
                .flatMapIterable { foodstuffs -> foodstuffs }
    }
}
